llvm/lib/Target/Mips/NanoMipsLoadStoreMultiple.cpp

```diff
@@ -0,0 +1,359 @@
+// lw a31, 8 (a9)
+// lw a16, 12(a9)
+void NMLoadStoreMultipleOpt::sortLoadStoreList(InstrList &LoadStoreList,
+                                               bool IsLoad) {
+  auto CompareInstructions = [this, IsLoad](MachineInstr *First,
+                                            MachineInstr *Second) {
+    Register FirstReg = First->getOperand(0).getReg();
+    Register SecondReg = Second->getOperand(0).getReg();
+    unsigned FirstRegNo = getRegNo(FirstReg);
+    unsigned SecondRegNo = getRegNo(SecondReg);
+
+    // For the zero register stores, sort instructions by the Offset.
+    if (!IsLoad && FirstRegNo == 0 && SecondRegNo == 0)
+      return First->getOperand(2).getImm() < Second->getOperand(2).getImm();
+    return FirstRegNo < SecondRegNo;
+  };
+  std::sort(LoadStoreList.begin(), LoadStoreList.end(), CompareInstructions);
+}
+
+// All instruction in the seqence should have the same Rs register, and
+// different Rt register.
+bool NMLoadStoreMultipleOpt::isValidLoadStore(MachineInstr &MI, bool IsLoad,
+                                              InstrList Sequence) {
+  unsigned Opcode = MI.getOpcode();
+  Register Rt, Rs;
+  if (IsLoad) {
+    // TODO: Handle unaligned loads and stores.
+    if (Opcode != Mips::LW_NM && Opcode != Mips::LWs9_NM)
+      return false;
+
+    Rt = MI.getOperand(0).getReg();
+    Rs = MI.getOperand(1).getReg();
+
+    // TODO: Rt and Rs can be equal, but only if that is the last load of
+    // the sequence.
+    if (Rt == Rs)
+      return false;
+
+  } else {
+    if (Opcode != Mips::SW_NM && Opcode != Mips::SWs9_NM)
+      return false;
+    Rt = MI.getOperand(0).getReg();
+    Rs = MI.getOperand(1).getReg();
+  }
+
+  if (Sequence.size() > 0) {
+    auto SeqRs = Sequence.back()->getOperand(1).getReg();
+    if (Rs != SeqRs)
+      return false;
+  }
+  auto RtExists = [&Rt](const MachineInstr *I) {
+    return I->getOperand(0).getReg() == Rt;
+  };
+  auto It = std::find_if(Sequence.begin(), Sequence.end(), RtExists);
+  // Zero register stores are a special case that does not require consequent
+  // $rt registers, but instead requires all $rt registers to be $zero.
+  if (It == Sequence.end() || getRegNo(Rt) == 0)
+    return true;
+  return false;
+}
+
+bool NMLoadStoreMultipleOpt::isValidNextLoadStore(LSIns Prev, LSIns Next,
+                                                  size_t &GapSize,
+                                                  size_t &CurrSeqSize,
+                                                  bool &RegGap) {
+  unsigned PrevRtNo = getRegNo(Prev.Rt);
+  unsigned DesiredRtNo = PrevRtNo != 0 ? (PrevRtNo + 1) : 0;
+  Register DesiredRtReg = RC.getRegister(DesiredRtNo);
+  if (Next.Offset == Prev.Offset + 4) {
+    // GAP, but offset ok
+    // lw a0, 8(a4)
+    // lw a1, 12(a4)
+    // lw a3, 16(a4)
+    if (Next.Rt != DesiredRtReg) {
+      // For now, the instruction like lw a3, 16(a4) insterupts the sequence.
+      if (CurrSeqSize < 2)
+        return false;
```

Comment:
What about sequences where we have, say,

```asm
lw a0, 8(a4)
lw a5, 12(a4)
lw a2, 16(a4)
```

where a sequence of 3 loads could be replaced by a single lwm and a move, but won't be detected due to making this check earlier when the sequence is only 1 long so far?


Edit: I guess that's not entirely relevant since there restriction that only the last in the sequence can require a move is more integrated below.